cgsmiles/drawing_utils.py

```diff
@@ -0,0 +1,193 @@
+    return edges, arom_edges, plain_edges
+
+def make_mapped_edges(graph, plain_edges):
+    """
+    Mapped edges are all those edges between atoms
+    that belong to the same fragment (i.e. where both
+    nodes have the same fragid attribute).
+    """
+    mapped_edges = defaultdict(list)
+    for idx, jdx in graph.edges:
+        if graph.nodes[idx]['fragid'] == graph.nodes[jdx]['fragid']:
+            fragid = graph.nodes[idx]['fragid'][0]
+            mapped_edges[fragid].extend(plain_edges[frozenset([idx, jdx])])
+        elif len(graph.nodes[idx]['fragid']) > 1 and\
+             graph.nodes[jdx]['fragid'][0] in graph.nodes[idx]['fragid']:
+            fragid = graph.nodes[jdx]['fragid'][0]
+            mapped_edges[fragid].extend(plain_edges[frozenset([idx, jdx])])
+        elif len(graph.nodes[jdx]['fragid']) > 1 and\
+             graph.nodes[idx]['fragid'][0] in graph.nodes[jdx]['fragid']:
+            fragid = graph.nodes[idx]['fragid'][0]
+            mapped_edges[fragid].extend(plain_edges[frozenset([idx, jdx])])
+    return mapped_edges
+
+def make_node_pies(graph, pos, cgmapping, colors, use_weights=False):
+    """
+    Generate the slices for the matplotlip pies used to draw nodes.
+    """
+    for idx, node in enumerate(graph.nodes):
+        position = pos[node]
+        fragids = graph.nodes[node].get('fragid', None)
+        if fragids and len(fragids) > 1:
+
+            # find the first fragid and compute the angle of the edge with z
+            neighbors = graph.neighbors(node)
+            pie_colors = []
+            edges = []
+            angles = []
+            for neigh in neighbors:
+                if graph.nodes[neigh]['fragid'][0] in fragids:
+                    edge = pos[neigh] - pos[node]
+                    edges.append(edge)
+                    angle = rotation_from_x_axis(edge)
+                    if angle < 0:
+                        angle += 360
+                    angles.append(angle)
+                    pie_colors.append(colors(graph.nodes[neigh]['fragid'][0]))
+
+            # compute the rotation to align the pie
+            startangle = angle_of_interest(edges[0], edges[1])
+            if startangle < 0:
+                startangle += 360
+            # sort angles and pie colors
+            pre_pie_colors = [x for _, x in sorted(zip(angles, pie_colors))][::-1]
+            angles = np.array(sorted(angles))
+            for angle in angles:
+                if angle < startangle: # and angle > startangle - (360/len(fragids)):
+                    color = pre_pie_colors.pop()
+                    pre_pie_colors = [color] + pre_pie_colors
+                    print(angle)
+            pie_colors = pre_pie_colors[::-1]
+            # here we set the weights that split the pies in equal slices
+            # I guess in principle on could also use the weight attribute
+            weight = 1/len(fragids)
+            slices = np.array([weight for n in range(0, len(fragids))])
```

Comment:
I have the feeling this gets significantly simpler if you create a list `relevant_neighbors` containing `(angle, node_idx, color)`. If you reverse sort that (reverse, because `plt.pie` is counterclockwise starting at the x-axis) and iterate you can populate `slices`? `startangle` is then something like `sorted_neighbors[0][0]/2`. Maybe.
Also, use weights; `plt.pie` normalizes

Reply:
Finally, maybe worth splitting out into a function that make a pie for just node n, and maybe separate the complex and easy cases.

Reply:
let's make this a maintenance thing